<!doctype html>
<html lang="en">
<head>
<meta charset="utf-8">

<title>Bio4j</title>

<meta name="description" content="Bio4j"/>
<meta name="apple-mobile-web-app-capable" content="yes" />
<meta name="apple-mobile-web-app-status-bar-style" content="black-translucent" />

<meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no">

<link rel="stylesheet" href="resources/reveal.js/css/reveal.min.css">
  <link rel="stylesheet" href="resources/reveal.js/css/theme/moon.css" id="theme">


<!-- For syntax highlighting -->
  <link rel="stylesheet" href="resources/reveal.js/lib/css/zenburn.css">


<!-- If the query includes 'print-pdf', use the PDF print sheet -->
<script>
  document.write( '<link rel="stylesheet" href="resources/reveal.js/css/print/' +
    ( window.location.search.match( /print-pdf/gi ) ? 'pdf' : 'paper' ) + 
    '.css" type="text/css" media="print">' );
</script>

<!--[if lt IE 9]>
<script src="resources/reveal.js/lib/js/html5shiv.js"></script>
<![endif]-->
</head>

<body>

<div class="reveal">

<!-- Any section element inside of this container is displayed as a slide -->
<div class="slides">

<section>
<h1>Bio4j</h1>
<h3><strong><a href="http://twitter.com/pablopareja">Pablo Pareja</a></strong></h3>
<p>
<h4><a href="http://graphdevroom.org"><strong>GraphDevRoom 2014</strong></a></h4>
</p>
</section>  


<section><section id="what-is-bio4j" class="titleslide slide level1"><h1>what is Bio4j</h1></section><section class="slide level2">

<h3 id="in-one-sentence">in one sentence</h3>
<p><strong>Bio4j</strong> is a bioinformatics <em>graph</em>-based data platform <strong>integrating</strong> most data available in the most representative <strong>open data sources</strong> around <strong>protein information</strong> available today.</p>
</section><section class="slide level2">

<h3 id="data">data</h3>
<ul>
<li><em><a href="http://www.uniprot.org/help/uniprotkb">UniProt KB</a></em> (SwissProt + Trembl)</li>
<li><em><a href="http://www.geneontology.org/">Gene Ontology</a></em> (GO)</li>
<li><em><a href="http://www.uniprot.org/help/uniref">UniRef</a></em> (50,90,100)</li>
<li><em><a href="http://www.ncbi.nlm.nih.gov/RefSeq/">RefSeq</a></em></li>
<li><em><a href="http://www.ncbi.nlm.nih.gov/Taxonomy/">NCBI taxonomy</a></em></li>
<li><em><a href="http://enzyme.expasy.org/">Expasy Enzyme DB</a></em></li>
</ul>
<!-- 4 or more '-' create a new slide -->

</section><section class="slide level2">

<h3 id="open">open!</h3>
<ul>
<li><strong>code</strong> <strong><a href="https://www.gnu.org/licenses/agpl-3.0.html">AGPLv3</a></strong></li>
<li><strong>data</strong> integrates only <strong><a href="http://okfn.org/opendata/">open data</a></strong></li>
<li><strong>implementation &amp; release</strong> process is 100% public and totally <strong>transparent</strong></li>
</ul>
</section></section>
<section><section id="why-bio4j" class="titleslide slide level1"><h1>why Bio4j?</h1></section><section class="slide level2">

<h3 id="biology-dbs-today">biology &amp; DBs today</h3>
<p>Highly <strong>interconnected</strong> overlapping knowledge <strong>spread</strong> through <em>different databases</em></p>
</section><section class="slide level2">

<figure>
<img src="resources/images/overlappingDBsBiology.jpg" />
</figure>
</section><section class="slide level2">

<h3 id="why-graphs">why graphs</h3>
<p>In most cases all data is modeled in <em><strong>Relational Databases</strong></em> or sometimes even just as plain <em><strong><code>CSV</code> files</strong></em>.</p>
</section><section class="slide level2">

<h3 id="why-graphs-1">why graphs</h3>
<p>That might be OK for simple scenarios but as the <strong>amount</strong> and <strong>diversity</strong> of <strong>data grows</strong>, <strong>domain models</strong> become crazily <strong>complicated</strong>!</p>
</section><section class="slide level2">

<p><em>Doesn’t look very compelling right? :)</em> <img src="resources/images/GOdomainModel.jpg" /></p>
</section><section class="slide level2">

<h3 id="why-graphs-2">why graphs</h3>
<p>With a relational paradigm the double implication</p>
<p><strong>Entity</strong> &lt;–&gt; <strong>Table</strong></p>
<p>does not go both ways</p>
</section><section class="slide level2">

<h3 id="not-so-good-implications">not-so-good implications</h3>
<ul>
<li><strong>Auxiliary tables</strong></li>
<li><strong>Artificial IDs</strong></li>
<li>Dealing with <strong>raw tables</strong> (in spite of Entity-relationship diagrams)</li>
<li><strong>Integrating</strong> new knowledge becomes <strong>difficult</strong></li>
</ul>
</section><section class="slide level2">

<h3 id="biology-table">biology <em>!=</em> table</h3>
<p><strong>Life</strong> in general and <strong>biology</strong> in particular are probably not 100% like a graph…</p>
<p>but one thing’s sure, they <em><strong>are not a set of tables!</strong></em></p>
</section><section class="slide level2">

<h3 id="why-graph-databases">why graph databases</h3>
<ul>
<li><strong>Data</strong> stored in a way that <strong>semantically represents its own structure</strong></li>
<li>Incorporating new data is easy –&gt; <strong>scalability</strong></li>
</ul>
</section><section class="slide level2">

<h3 id="why-graph-databases-1">why graph databases</h3>
<ul>
<li><strong>Vertex-centric</strong> <em>(local)</em> indices allow for complex traversals –&gt; overcoming <strong>supernode problem</strong></li>
</ul>
</section><section class="slide level2">

<h3 id="cloud">cloud</h3>
<ul>
<li>data as a service</li>
<li>machine configurations</li>
</ul>
</section></section>
<section><section id="details-about-bio4j" class="titleslide slide level1"><h1>details about Bio4j</h1></section><section class="slide level2">

<h3 id="a-bit-of-history">a bit of history</h3>
<p>From the beginnings to the BigData platform it is today</p>
</section><section class="slide level2">

<h3 id="how-it-all-started">how it all started</h3>
<ul>
<li>Need for <strong>massive access</strong> to <em>Gene Ontology</em> annotations</li>
<li><a href="http://bg7.ohnosequences.com/"><strong>BG7</strong></a> bacterial genome annotation system</li>
<li>Need for massive direct access to <strong>protein information</strong></li>
</ul>
</section><section class="slide level2">

<h3 id="more-and-more-data">more and more data!</h3>
<ul>
<li>As <em>other</em> data sources were becoming a <em>bottleneck</em> they were being added to Bio4j</li>
<li>First it was Uniprot KB, then Uniref and <strong>we didn’t stop yet :)</strong></li>
</ul>
</section><section class="slide level2">

<h3 id="numbers">numbers</h3>
<ul>
<li><span class="math">10<sup>9</sup></span> edges</li>
<li><span class="math">2 × 10<sup>8</sup></span> nodes</li>
<li><span class="math">6 × 10<sup>8</sup></span> properties</li>
<li><span class="math">150</span> edge types</li>
<li><span class="math">40</span> node types</li>
</ul>
<!-- * **`1,216,993,547`** relationships
* `190,625,351` nodes
* `584,436,429` properties
* `144` relationship types
* `42` node types
 -->
</section><section class="slide level2">

<h3 id="bio4j-structure">Bio4j structure</h3>
<p>Bio4j importing process is <strong>modular</strong> and <strong>customizable</strong> allowing you to import just the data you are interested in.</p>
</section><section class="slide level2">

<h3 id="data-sources---modules-i">data sources - modules I</h3>
<ul>
<li class="fragment">Gene Ontology (GO)</li>
<li class="fragment">ExPASy Enzyme DB</li>
<li class="fragment">RefSeq</li>
</ul>
</section><section class="slide level2">

<h3 id="data-sources---modules-ii">data sources - modules II</h3>
<ul>
<li class="fragment">UniRef -&gt; 50, 90, 100</li>
<li class="fragment">NCBI taxonomy tree -&gt; GI index</li>
<li class="fragment">Uniprot KB -&gt; Swissprot/Trembl, interactions…</li>
</ul>
</section><section class="slide level2">

<h3 id="data-sources---modules-iii">data sources - modules III</h3>
<p>Just keep in mind that you must be <strong>coherent</strong></p>
<p><em>e.g. you cannot import protein interactions if you didn’t import any protein yet!</em></p>
</section><section class="slide level2">

<h3 id="bio4j-apis">Bio4j APIs</h3>
<ol type="1">
<li class="fragment">abstract <strong>domain model</strong></li>
<li class="fragment"><strong>Blueprints</strong> implementation</li>
<li class="fragment"><strong>technology-specific</strong> versions</li>
</ol>
</section><section class="slide level2">

<h3 id="domain-model">domain model</h3>
<p>Bio4j database has a <strong>well-defined</strong> domain model and all nodes and relationships comply with this abstract model</p>
</section><section class="slide level2">

<figure>
<img src="resources/images/Bio4jDomainModelWithCardinality.jpg" />
</figure>
</section><section class="slide level2">

<h3 id="why-domain-model">why domain model?</h3>
<ul>
<li>abstract over Blueprints</li>
<li>more precise <strong>typing</strong></li>
<li>implementations can use technology-specific features</li>
</ul>
</section><section class="slide level2">

<h3 id="key-advantage">key advantage</h3>
<p><em>Different</em> <strong>graph topologies</strong> at the storage level, <em>same</em> <strong>domain model</strong>.</p>
<p>Example: use <strong>type nodes</strong> in <em>Titan</em>, <strong>labels</strong> in <em>Neo4j</em>.</p>
</section><section class="slide level2">

<h3 id="blueprints-layer">blueprints layer</h3>
<p>A default <strong><a href="https://github.com/tinkerpop/blueprints/">Blueprints</a></strong> implementation of the abstract model.</p>
<p>Apart from the set of interfaces developed as the <strong>first layer</strong> for the <em>domain model</em> there’s an <strong>extra layer</strong> that uses <em>Blueprints</em>. This way we’re going one step further for making the domain model <strong>independent</strong> from the choice of <em>database technology</em></p>
</section><section class="slide level2">

<h3 id="technology-specific">technology-specific</h3>
<p>Optimizations, features, etc.</p>
<ul>
<li><strong><a href="https://github.com/neo4j/neo4j">Neo4j</a></strong></li>
<li><strong><a href="https://github.com/thinkaurelius/titan/">Titan</a></strong> (WIP)</li>
<li><strong><a href="https://github.com/orientechnologies/orientdb/">OrientDB</a></strong> (planned)</li>
</ul>
</section><section class="slide level2">

<h3 id="why-neo4j">why Neo4j</h3>
<ul>
<li class="fragment">wide <strong>adoption</strong></li>
<li class="fragment">stable</li>
<li class="fragment"><strong>Cypher</strong></li>
</ul>
</section><section class="slide level2">

<h3 id="why-titan">why Titan</h3>
<ul>
<li class="fragment"><strong>local!</strong> indexes</li>
<li class="fragment"><strong>on-disk</strong> access</li>
<li class="fragment"><strong>type</strong> definitions -&gt; <em>constraints!</em></li>
</ul>
</section><section class="slide level2">

<h3 id="bio4j-and-the-cloud">Bio4j and the cloud</h3>
<ul>
<li><strong>Interoperability</strong> and data distribution</li>
<li><strong>Backup</strong> and <strong>storage</strong></li>
<li><strong>Scalability</strong></li>
<li>Applications and service providers on the cloud</li>
<li>Cost-effective</li>
</ul>
</section><section class="slide level2">

<h3 id="dev-and-release-process">dev and release process</h3>
<ul>
<li>coordinate <strong>data</strong> and <strong>code</strong></li>
<li><strong><a href="http://semver.org/spec/v2.0.0.html">Semantic Versioning</a></strong></li>
<li><strong>Cloud</strong> integration, distribution, deployment, …</li>
</ul>
</section><section class="slide level2">

<h3 id="how">how?</h3>
<ul>
<li><strong><a href="http://ohnosequences/statika">Statika</a></strong> cloud, data + code, modules (see <a href="https://fosdem.org/2014/schedule/event/graphdevroom_bio4j_1/">next talk</a>)</li>
<li><strong><a href="https://github.com/sbt/sbt">sbt</a></strong> build Java + Scala, automated Bio4j-specific test &amp; release</li>
<li><strong><a href="https://github.com/bio4j">git + github</a></strong> versioning, docs, collaboration, coordination</li>
</ul>
</section></section>
<section><section id="how-to-use-bio4j" class="titleslide slide level1"><h1>how to use Bio4j?</h1></section><section class="slide level2">

<h3 id="how-we-use-it">how we use it</h3>
<ul>
<li><strong><a href="http://bg7.ohnosequences.com">bg7</a></strong> genome annotation</li>
<li><strong>mg7</strong> metagenomics analysis</li>
<li>comparative genomics, network analysis, genome assembly, …</li>
</ul>
</section><section class="slide level2">

<h3 id="case-study-ii">case study II</h3>
<p><strong>Ohio State University</strong></p>
<ul>
<li><strong>Integration</strong> and <strong>analysis</strong> of Chip-seq data</li>
<li><strong>Modeling</strong> genomic information and <strong>gene regulatory networks</strong></li>
</ul>
</section><section class="slide level2">

<h3 id="case-study-iii">case study III</h3>
<p><strong>Berkeley Phylogenomics Group</strong></p>
<ul>
<li>Graph database for <em>Big Data challenges</em> in <strong>genomics</strong> developed <strong>on top of Bio4j</strong></li>
</ul>
</section><section class="slide level2">

<h3 id="community">community</h3>
<ul>
<li><strong><a href="http://twitter.com/bio4j">@bio4j</a></strong> twitter</li>
<li><strong><a href="https://github.com/bio4j">bio4j</a></strong> github org</li>
<li><strong><a href="http://groups.google.com/group/bio4j-user">bio4j-user</a></strong> google group</li>
<li><strong><a href="http://www.linkedin.com/groups/Bio4j-3890937">bio4j</a></strong> linkedin</li>
</ul>
</section></section>
<section><section id="whos-doing-bio4j" class="titleslide slide level1"><h1>who’s doing Bio4j?</h1></section><section class="slide level2">

<h3 id="oh-no-sequences">oh no sequences!</h3>
<p><strong><a href="http://era7bioinformatics.com">Era7 bioinformatics</a></strong> R&amp;D group</p>
<ul>
<li><strong>web</strong> -&gt; <a href="http://ohnosequences.com">ohnosequences.com</a></li>
<li><strong>Github</strong> -&gt; <a href="https://github.com/ohnosequences">ohnosequences</a></li>
</ul>
</section><section class="slide level2">

<h3 id="team">team</h3>
<ul>
<li><strong><a href="http://ohnosequences.com/ppareja">Pablo Pareja</a></strong> <br /> project leader &amp; main dev</li>
<li><strong><a href="http://ohnosequences.com/eparejatobes">Eduardo Pareja-Tobes</a></strong><br /> technology &amp; architecture</li>
<li><strong><a href="http://ohnosequences.com/rtobes">Raquel Tobes</a></strong> <br /> bio data integration</li>
</ul>
</section><section class="slide level2">

<h3 id="team-1">team</h3>
<ul>
<li><strong><a href="http://ohnosequences.com/aalekhin">Alexey Alekhin</a></strong> <br /> Statika, release process, dev</li>
<li><strong><a href="http://ohnosequences.com/mmanrique">Marina Manrique</a></strong> <br /> bio data integration</li>
<li><strong><a href="http://ohnosequences.com/ekovach">Evdokim Kovach</a></strong><br /> dev</li>
</ul>
</section></section>

</div>

<script src="resources/reveal.js/lib/js/head.min.js"></script>
<script src="resources/reveal.js/js/reveal.min.js"></script>

<script>
  // Full list of configuration options available here:
  // https://github.com/hakimel/reveal.js#configuration
  Reveal.initialize({
    controls: true,
    progress: true,
    history: true,
    center: true,

  // available themes are in /resources/reveal.js/css/theme
      theme: Reveal.getQueryHash().theme || 'moon', 
  
  // default/cube/page/concave/zoom/linear/fade/none
      transition: Reveal.getQueryHash().transition || 'linear',
  
  // Optional libraries used to extend on reveal.js
  dependencies: [
    { src: 'resources/reveal.js/lib/js/classList.js', condition: function() { return !document.body.classList; } },
    { src: 'resources/reveal.js/plugin/markdown/marked.js', condition: function() { return !!document.querySelector( '[data-markdown]' ); } },
    { src: 'resources/reveal.js/plugin/markdown/markdown.js', condition: function() { return !!document.querySelector( '[data-markdown]' ); } },
    { src: 'resources/reveal.js/plugin/highlight/highlight.js', async: true, callback: function() { hljs.initHighlightingOnLoad(); } },
    { src: 'resources/reveal.js/plugin/zoom-js/zoom.js', async: true, condition: function() { return !!document.body.classList; } },
    { src: 'resources/reveal.js/plugin/notes/notes.js', async: true, condition: function() { return !!document.body.classList; } }
    // { src: 'resources/reveal.js/plugin/remotes/remotes.js', async: true, condition: function() { return !!document.body.classList; } }
  ]
  });

</script>

</body>
</html>
